google-cloud-pubsub/lib/google/cloud/pubsub/async_publisher.rb

```diff
@@ -224,40 +282,93 @@ def run_background
           end
         end
 
-        def publish_batch!
-          return unless @batch
+        def resolve_batch_for_message msg
+          @batches[msg.ordering_key] ||= Batch.new self, msg.ordering_key
+        end
 
-          publish_batch_async @topic_name, @batch
-          @batch = nil
-          @first_published_at = nil
+        def resolve_batch_for_ordering_key ordering_key
+          @batches[ordering_key]
+        end
+
+        def publish_batches! stop: nil
+          @batches.reject! { |_ordering_key, batch| batch.empty? }
+          @batches.values.each do |batch|
+            ready = batch.publish! stop: stop
+            publish_batch_async @topic_name, batch if ready
+          end
+          # Set published_at to nil to wait indefinitely
+          @published_at = nil
         end
 
         def publish_batch_async topic_name, batch
+          # TODO: raise unless @publish_thread_pool.running?
           return unless @publish_thread_pool.running?
 
           Concurrent::Promises.future_on(
             @publish_thread_pool, topic_name, batch
-          ) do |t_name, btch|
-            publish_batch_sync t_name, btch
-          end
+          ) { |t, b| publish_batch_sync t, b }
         end
 
+        # rubocop:disable Metrics/AbcSize
+        # rubocop:disable Metrics/MethodLength
+
         def publish_batch_sync topic_name, batch
-          grpc = @service.publish topic_name, batch.messages
-          batch.items.zip Array(grpc.message_ids) do |item, id|
-            next unless item.callback
+          # The only batch methods that are safe to call from the loop are
+          # rebalance! and flush! because they are the only methods that are
```

Comment:
This comment seems inaccurate. `Batch#reset!` is called in the loop, and it is synchronized.

Reply:
Yep, this was an early comment before many, many refactorings. Corrected.